retrofit2/src/main/java/com/linecorp/armeria/client/http/retrofit2/ArmeriaRetrofitBuilder.java

```diff
@@ -0,0 +1,221 @@
+ * {@link HttpUrl} throwing exception when parsing http://group:foo . To avoid your client have possibility to
+ * access domain which starts with __group__, you can use {@link ArmeriaRetrofitBuilder#groupPrefix(String)} to
+ * customize the internal behavior.
+ *
+ * <p>If you want to decorate HttpClient, using {@link ArmeriaRetrofitBuilder#newClientFunction(Function)} to
+ * customize.
+ * <pre>{@code
+ *
+ * Retrofit retrofit = new ArmeriaRetrofitBuilder("none+http://localhost:8080/")
+ *     .newClientFunction(uri -> new ClientBuilder(uri)
+ *             .decorator(HttpRequest.class, HttpResponse.class, LoggingClient::new)
+ *             .build(HttpClient.class))
+ *     .build();
+ * }
+ * </pre>
+ */
+public final class ArmeriaRetrofitBuilder {
+
+    private static final Pattern GROUP_PREFIX_PATTERN = Pattern.compile("^[_0-9a-z]+$");
+    private static final String DEFAULT_GROUP_PREFIX = "__group__";
+    private static final Function<String, HttpClient> DEFAULT_NEW_CLIENT_FUNCTION =
+            url -> Clients.newClient(ClientFactory.DEFAULT, url, HttpClient.class);
+    public static final String SLASH = "/";
+
+    private final Retrofit.Builder retrofitBuilder;
+    private String baseUrl;
+    private HttpClient baseHttpClient;
+    private String basePath;
+    private String groupPrefix = DEFAULT_GROUP_PREFIX;
+    private Function<String, HttpClient> newClientFunction = DEFAULT_NEW_CLIENT_FUNCTION;
+
+    /**
+     * Creates a {@link Retrofit.Builder} with {@link ArmeriaCallFactory} using the specified
+     * url.
+     *
+     * @see Builder#baseUrl(String)
+     */
+    public ArmeriaRetrofitBuilder(String baseUrl) {
+        String path = URI.create(baseUrl).getPath();
+        if (!path.isEmpty() && !SLASH.equals(path.substring(path.length() - 1))) {
+            throw new IllegalArgumentException("baseUrl must end in /: " + baseUrl);
+        }
+        this.baseUrl = baseUrl;
+        retrofitBuilder = new Retrofit.Builder();
+    }
+
+    /**
+     * Creates a {@link Retrofit.Builder} with {@link ArmeriaCallFactory} using the specified
+     * {@link HttpClient} instance.
+     */
+    public ArmeriaRetrofitBuilder(HttpClient baseHttpClient) {
+        this(baseHttpClient, SLASH);
+    }
+
+    /**
+     * Creates a {@link Retrofit.Builder} with {@link ArmeriaCallFactory} using the specified
+     * {@link HttpClient} instance.
+     */
+    public ArmeriaRetrofitBuilder(HttpClient baseHttpClient, String basePath) {
+        final String path = baseHttpClient.uri().getPath();
+        if (!path.isEmpty() && !SLASH.equals(path)) {
+            throw new IllegalArgumentException(
+                    "ArmeriaRetrofitBuilder doesn't support http client's uri contains any path element," +
+                    " current path: " + path +
+                    ". If you want to using uri with path, please using constructor with basePath argument.");
```

Comment:
How about:

    "baseHttpClient should not have any path element in its URI: " + path +
    " (Use the constructor that accepts 'basePath' instead to specify a URI with a path.)"